### ISSUE TITLE
Running rr in qemu VM on Zen not working

### ISSUE THREAD

**Issue by bernhardu** [rr-debugger/rr#2706]
I hope this is the right place to ask this question?
On my `Ryzen 7 1700` I have started a qemu VM with `-enable-kvm -cpu host -smp 16`,
running an debian testing as host and guest, both with `nospec_store_bypass_disable` under a `kernel 5.8.10-1`.
Does someone got this running?

```
benutzer@debian:~/obj$ bin/rr /bin/ls
On Zen CPUs, rr will not work reliably unless you disable the hardware SpecLockMap optimization.
For instructions on how to do this, see https://github.com/mozilla/rr/wiki/Zen
rr: Saving execution to trace directory `/home/benutzer/.local/share/rr/ls-1'.
32                             CMakeCache.txt...
```

```
[Mi Sep 30 15:04:16 2020] unchecked MSR access error: WRMSR to 0xc0010200 (tried to write 0x00000200001100d1) at rIP: 0xffffffff83864774 (native_write_msr+0x4/0x20)
[Mi Sep 30 15:04:16 2020] Call Trace:
[Mi Sep 30 15:04:16 2020]  amd_pmu_disable_event+0x22/0x90
[Mi Sep 30 15:04:16 2020]  x86_pmu_stop+0x3b/0xa0
[Mi Sep 30 15:04:16 2020]  __perf_event_period+0x81/0xf0
[Mi Sep 30 15:04:16 2020]  event_function+0x8a/0xd0
[Mi Sep 30 15:04:16 2020]  ? perf_log_throttle+0x120/0x120
[Mi Sep 30 15:04:16 2020]  remote_function+0x3e/0x50
[Mi Sep 30 15:04:16 2020]  generic_exec_single+0xa0/0xd0
[Mi Sep 30 15:04:16 2020]  ? do_softirq_own_stack+0x39/0x50
[Mi Sep 30 15:04:16 2020]  smp_call_function_single+0xd6/0x120
[Mi Sep 30 15:04:16 2020]  ? sw_perf_event_destroy+0x60/0x60
[Mi Sep 30 15:04:16 2020]  event_function_call+0x97/0x170
[Mi Sep 30 15:04:16 2020]  ? perf_log_throttle+0x120/0x120
[Mi Sep 30 15:04:16 2020]  ? alloc_perf_context+0x110/0x110
[Mi Sep 30 15:04:16 2020]  _perf_event_period+0x69/0x90
[Mi Sep 30 15:04:16 2020]  _perf_ioctl+0x311/0xab0
[Mi Sep 30 15:04:16 2020]  ? _cond_resched+0x16/0x40
[Mi Sep 30 15:04:16 2020]  perf_ioctl+0x3d/0x60
[Mi Sep 30 15:04:16 2020]  ksys_ioctl+0x82/0xc0
[Mi Sep 30 15:04:16 2020]  __x64_sys_ioctl+0x16/0x20
[Mi Sep 30 15:04:16 2020]  do_syscall_64+0x44/0xc0
[Mi Sep 30 15:04:16 2020]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[Mi Sep 30 15:04:16 2020] RIP: 0033:0x7f07c1b82d87
[Mi Sep 30 15:04:16 2020] Code: 00 00 00 48 8b 05 09 91 0c 00 64 c7 00 26 00 00 00 48 c7 c0 ff ff ff ff c3 66 2e 0f 1f 84 00 00 00 00 00 b8 10 00 00 00 0f 05 <48> 3d 01 f0 ff ff 73 01 c3 48 8b 0d d9 90 0c 00 f7 d8 64 89 01 48
[Mi Sep 30 15:04:16 2020] RSP: 002b:00007ffc2e573428 EFLAGS: 00000206 ORIG_RAX: 0000000000000010
[Mi Sep 30 15:04:16 2020] RAX: ffffffffffffffda RBX: 0000000000000000 RCX: 00007f07c1b82d87
[Mi Sep 30 15:04:16 2020] RDX: 00007ffc2e573440 RSI: 0000000040082404 RDI: 0000000000000003
[Mi Sep 30 15:04:16 2020] RBP: 00007ffc2e5734f0 R08: 0000000000000008 R09: 0000000000000000
[Mi Sep 30 15:04:16 2020] R10: fffffffffffff247 R11: 0000000000000206 R12: 0000000000000000
[Mi Sep 30 15:04:16 2020] R13: 0000000000000000 R14: 0000000000000000 R15: 0000000000000000
```

```
root@debian:~# grep Spec /proc/self/status
Speculation_Store_Bypass:       vulnerable
root@debian:~# /home/benutzer/rr/scripts/zen_workaround.py
Traceback (most recent call last):
  File "/home/benutzer/rr/scripts/zen_workaround.py", line 35, in <module>
    val = read_msr(cpu)
  File "/home/benutzer/rr/scripts/zen_workaround.py", line 27, in read_msr
    (val,) = struct.unpack('<q', os.read(msr, 8))
OSError: [Errno 5] Input/output error
```

And replay fails this way:
```
benutzer@debian:~/obj$ bin/rr replay -a $HOME/.local/share/rr/latest-trace
On Zen CPUs, rr will not work reliably unless you disable the hardware SpecLockMap optimization.
For instructions on how to do this, see https://github.com/mozilla/rr/wiki/Zen

rr: Warning: You appear to be running in a VMWare guest with a bug
    where a conditional branch instruction between two CPUID instructions
    sometimes fails to be counted by the conditional branch performance
    counter. Work around this problem by adding
        monitor_control.disable_hvsim_clusters = true
    to your .vmx file.

[FATAL /home/benutzer/rr/src/ReplaySession.cc:1078:check_ticks_consistency()] 
 (task 672 (rec:667) at time 15)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'INSTRUCTION_TRAP'; expected 111, got 121
...
```

**Comment by khuey** [comment]
Maybe @glandium has some idea, but getting rr working in virtualization is a fair bit harder than just bare metal.

**Comment by glandium** [comment]
Did you run the zen workaround on the host or in the guest?

**Comment by bernhardu** [comment]
I booted the host with kernel parameter `nospec_store_bypass_disable`.
Then on the host zen_workaround.py returns `Zen workaround in place`.

Then booted the guest VM with kernel parameter `nospec_store_bypass_disable`.
There zen_workaround.py returns the error shown in my first post.

**Comment by glandium** [comment]
Can you try the workaround as a kernel module? https://github.com/mozilla/rr/wiki/Zen#kernel-module (on the host)

**Comment by bernhardu** [comment]
Yes, I removed the kernel parameter, rebooted and loaded the module.
(Running the tests on the host worked fine like before. 4 tests need a second attempt.)

Inside the VM a difference was that when recording or replaying the message "On Zen CPUs, rr will not work..." was not written.
Unfortunately all other messages and errors were the same.

I wondered if the issue relates to having `unchecked MSR access error` using `0xc0010200` ([MSR_F15H_PERF_CTL](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/x86/include/asm/msr-index.h#n489)),
but e.g. zen_workaround.py uses `0xc0011020` ([MSR_AMD64_LS_CFG](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/x86/include/asm/msr-index.h#n448)). I steppind into the recording rr and found that the `unchecked MSR access` gets triggered inside the `check_for_ioc_period_bug`.

**Comment by glandium** [comment]
> Inside the VM a difference was that when recording or replaying the message "On Zen CPUs, rr will not work..." was not written.

I think that's all that's important.

**Comment by bernhardu** [comment]
> > Inside the VM a difference was that when recording or replaying the message "On Zen CPUs, rr will not work..." was not written.
> 
> I think that's all that's important.

I am not sure if I understand you right here.
Yes, the kernel module is an improvement over the kernel parameter,
but replaying still fails with `ticks mismatch for 'INSTRUCTION_TRAP'` with the warning being a VMWare guest?

Can you reproduce this behaviour, or is my setup are my qemu parameters faulty?

**Comment by glandium** [comment]
Yeah, I can reproduce. It doesn't seem related to that kernel message, though.

**Comment by KaneTW** [comment]
Was this ever resolved? Running into the same warning message as OP, Ryzen 3900X with a KVM/qemu host. Zen patch successfully applies on host, fails on guest. Zen error message doesn't appear.

**Comment by rocallahan** [comment]
Seems most likely to me that KVM's PMU virtualization is not fully accurate on Zen :-(.

Try recording something simple on the host that you can replay correctly there, and try replaying it in the guest, and let us know what happens.

**Comment by KaneTW** [comment]
`rr replay -a` works correctly on host, but fails in the VM with
```
./SimulaDev/result/bin/rr replay -a who-0/

rr: Warning: You appear to be running in a VMWare guest with a bug
    where a conditional branch instruction between two CPUID instructions
    sometimes fails to be counted by the conditional branch performance
    counter. Work around this problem by adding
        monitor_control.disable_hvsim_clusters = true
    to your .vmx file.

[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /usr/bin/who changed: replay divergence likely, but continuing anyway. inode: 12853626/81398684; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 55912/55912; mtime: 1613428985/1606497679
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /usr/bin/who changed: replay divergence likely, but continuing anyway. inode: 12853626/81398684; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 55912/55912; mtime: 1613428985/1606497679
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /usr/bin/who changed: replay divergence likely, but continuing anyway. inode: 12853626/81398684; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 55912/55912; mtime: 1613428985/1606497679
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /usr/bin/who changed: replay divergence likely, but continuing anyway. inode: 12853626/81398684; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 55912/55912; mtime: 1613428985/1606497679
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /lib64/ld-2.32.so changed: replay divergence likely, but continuing anyway. inode: 12853628/95556651; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 177528/177528; mtime: 1613428985/1613426290
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /lib64/ld-2.32.so changed: replay divergence likely, but continuing anyway. inode: 12853628/95556651; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 177528/177528; mtime: 1613428985/1613426290
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /lib64/ld-2.32.so changed: replay divergence likely, but continuing anyway. inode: 12853628/95556651; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 177528/177528; mtime: 1613428985/1613426290
[ERROR /build/source/src/TraceStream.cc:1122:read_mapped_region()] Metadata of /lib64/ld-2.32.so changed: replay divergence likely, but continuing anyway. inode: 12853628/95556651; mode: 33261/33261; uid: 1000/0; gid: 1000/0; size: 177528/177528; mtime: 1613428985/1613426290
[FATAL /build/source/src/ReplaySession.cc:1082:check_ticks_consistency()] 
 (task 1903 (rec:6266) at time 15)
 -> Assertion `ticks_now == trace_ticks' failed to hold. ticks mismatch for 'INSTRUCTION_TRAP'; expected 0, got 435
Tail of trace dump:
=== Start rr backtrace:
./SimulaDev/result/bin/rr(_ZN2rr13dump_rr_stackEv+0x44)[0x68f224]
./SimulaDev/result/bin/rr(_ZN2rr9GdbServer15emergency_debugEPNS_4TaskE+0x1a2)[0x4ce218]
./SimulaDev/result/bin/rr[0x4ff6dc]
./SimulaDev/result/bin/rr(_ZN2rr21EmergencyDebugOstreamD1Ev+0x62)[0x4ff936]
./SimulaDev/result/bin/rr(_ZN2rr13ReplaySession23check_ticks_consistencyEPNS_10ReplayTaskERKNS_5EventE+0x128)[0x5e3344]
./SimulaDev/result/bin/rr(_ZN2rr13ReplaySession28emulate_deterministic_signalEPNS_10ReplayTaskEiRKNS0_15StepConstraintsE+0x37d)[0x5e3729]
./SimulaDev/result/bin/rr(_ZN2rr13ReplaySession18try_one_trace_stepEPNS_10ReplayTaskERKNS0_15StepConstraintsE+0x130)[0x5e557a]
./SimulaDev/result/bin/rr(_ZN2rr13ReplaySession11replay_stepERKNS0_15StepConstraintsE+0x17b)[0x5e69e7]
./SimulaDev/result/bin/rr[0x5dabf1]
./SimulaDev/result/bin/rr[0x5dbe13]
./SimulaDev/result/bin/rr[0x5dc30f]
./SimulaDev/result/bin/rr(_ZN2rr13ReplayCommand3runERSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS7_EE+0x414)[0x5dd182]
./SimulaDev/result/bin/rr(main+0x227)[0x6aa5af]
/nix/store/9df65igwjmf2wbw0gbrrgair6piqjgmi-glibc-2.31/lib/libc.so.6(__libc_start_main+0xed)[0x7fd20f58dc7d]
./SimulaDev/result/bin/rr(_start+0x2a)[0x43055a]
=== End rr backtrace
Launch gdb with
  gdb '-l' '10000' '-ex' 'set sysroot /' '-ex' 'target extended-remote 127.0.0.1:1903' /home/david/who-0/mmap_hardlink_3_who

```
Trace available at https://kane.cx/~kane/who-0.tar.xz

I can provide remote access to a separate VM if that makes debugging easier.

**Comment by rocallahan** [comment]
435 ticks when we expected zero is pretty bad!

I wonder if it's counting ticks in the guest kernel even though we ask it not to.

**Comment by rocallahan** [comment]
Maybe write a small program that does a raw syscall in a loop, and try to use `perf` to see if it's counting events in the guest kernel  as well as userspace?

**Comment by KaneTW** [comment]
Ok, I'll try that. The previous run was:
* `rr record who` on the host (not sure if VM was on or not at this point, actually)
* successful `rr replay -a` on the host
* failed `rr replay -a` on the guest

Should I repeat that but with 
* `perf record <test program>` on the host while VM is running
* `perf report` on the host?

**Comment by rocallahan** [comment]
No.

My wild-guess hypothesis is that when running `perf` or using `rr` in the guest, events in the guest kernel are being counted even though they shouldn't be.

**Comment by rocallahan** [comment]
But I'm not sure what the easiest way is to test that hypothesis.

**Comment by rocallahan** [comment]
Actually I guess the easiest way would be to compare the results of
```
perf stat -e branches /bin/ls /
perf stat -e branches --all-user /bin/ls /
perf stat -e branches --all-kernel /bin/ls /
```
in the guest. Basically the latter two numbers should roughly add up to the first number. If my hypothesis is correct then the first number would roughly equal the second number.

**Comment by KaneTW** [comment]
```
root@simula-dev-vm:~# perf stat -e branches /bin/ls /
bin   cdrom  etc   lib	  lib64   lost+found  mnt  opt	 root  sbin  srv       sys  usr
boot  dev    home  lib32  libx32  media       nix  proc  run   snap  swapfile  tmp  var

 Performance counter stats for '/bin/ls /':

           333,194      branches                                                    

       0.006604420 seconds time elapsed

       0.000000000 seconds user
       0.006681000 seconds sys


root@simula-dev-vm:~# perf stat -e branches --all-user /bin/ls /
bin   cdrom  etc   lib	  lib64   lost+found  mnt  opt	 root  sbin  srv       sys  usr
boot  dev    home  lib32  libx32  media       nix  proc  run   snap  swapfile  tmp  var

 Performance counter stats for '/bin/ls /':

           102,222      branches                                                    

       0.000852996 seconds time elapsed

       0.000916000 seconds user
       0.000000000 seconds sys


root@simula-dev-vm:~# perf stat -e branches --all-kernel /bin/ls /
bin   cdrom  etc   lib	  lib64   lost+found  mnt  opt	 root  sbin  srv       sys  usr
boot  dev    home  lib32  libx32  media       nix  proc  run   snap  swapfile  tmp  var

 Performance counter stats for '/bin/ls /':

           223,305      branches                                                    

       0.009991108 seconds time elapsed

       0.000000000 seconds user
       0.010083000 seconds sys
```

Seems as expected.

**Comment by KaneTW** [comment]
I've ran the following program on the host and guest and got pretty different results:

Host (perf version 5.11):
```
# Total Lost Samples: 0
#
# Samples: 8  of event 'cycles'
# Event count (approx.): 2651838
#
# Overhead  Command  Shared Object     Symbol
# ........  .......  ................  ................................
#
    89.46%  syscall  [kernel.vmlinux]  [k] enqueue_hrtimer
    10.13%  syscall  [kernel.vmlinux]  [k] get_page_from_freelist
     0.39%  syscall  [kernel.vmlinux]  [k] _extract_crng
     0.01%  perf     [kernel.vmlinux]  [k] __memcpy
     0.00%  perf     [kernel.vmlinux]  [k] perf_event_exec
     0.00%  perf     [kernel.vmlinux]  [k] perf_event_addr_filters_exec
```
Guest (perf version 5.8.18):
```
Samples: 45  of event 'cycles', Event count (approx.): 579737
Overhead  Command  Shared Object      Symbol
  15.69%  syscall  [kernel.kallsyms]  [k] __cgroup_account_cputime
  10.20%  syscall  libc-2.31.so       [.] __nanosleep
   6.96%  syscall  [kernel.kallsyms]  [k] _copy_from_user
   6.96%  syscall  [kernel.kallsyms]  [k] pvclock_clocksource_read
   6.07%  syscall  [kernel.kallsyms]  [k] __hrtimer_init
   6.07%  syscall  [kernel.kallsyms]  [k] copy_user_generic_string
   6.07%  syscall  [kernel.kallsyms]  [k] schedule
   5.96%  syscall  libc-2.31.so       [.] clock_nanosleep@@GLIBC_2.17
   4.08%  syscall  [kernel.kallsyms]  [k] entry_SYSCALL_64
   3.92%  syscall  [kernel.kallsyms]  [k] acct_collect
   3.92%  syscall  [kernel.kallsyms]  [k] ctx_sched_out
   3.90%  syscall  [kernel.kallsyms]  [k] update_blocked_averages
   3.03%  syscall  [kernel.kallsyms]  [k] __schedule
   3.03%  syscall  [kernel.kallsyms]  [k] finish_task_switch
   3.03%  syscall  [kernel.kallsyms]  [k] rb_insert_color
   2.98%  syscall  [kernel.kallsyms]  [k] common_nsleep
   2.04%  syscall  [kernel.kallsyms]  [k] account_process_tick
   2.04%  syscall  syscall            [.] _init
   2.04%  syscall  syscall            [.] main
   2.02%  syscall  [kernel.kallsyms]  [k] __hrtimer_run_queues
   0.00%  syscall  [kernel.kallsyms]  [k] do_nanosleep
   0.00%  perf     [kernel.kallsyms]  [k] strrchr
   0.00%  syscall  [kernel.kallsyms]  [k] __x64_sys_clock_nanosleep
   0.00%  syscall  [kernel.kallsyms]  [k] do_syscall_64
   0.00%  syscall  [kernel.kallsyms]  [k] entry_SYSCALL_64_after_hwframe
   0.00%  syscall  [kernel.kallsyms]  [k] native_write_msr
   0.00%  syscall  [kernel.kallsyms]  [k] switch_fpu_return
```
Program:
```C
#include <time.h>

const struct timespec req  = { .tv_sec=0, .tv_nsec=1000*1000*100 };

int main() {
  struct timespec rem;
  while(1) {
    nanosleep(&req, &rem);
  }
}
```

**Comment by rocallahan** [comment]
> Seems as expected.

Yes.

I'm not sure what to conclude from those profiles.

**Comment by KaneTW** [comment]
I find it a bit weird that you actually see syscalls to nanosleep in the guest, but not on the host.

**Comment by khuey** [comment]
The root cause of this is probably https://lkml.org/lkml/2022/5/3/21

**Comment by khuey** [comment]
Yeah I'm confident that's the fix for the performance counter issues here.  `rr record` and `rr replay -a` should work fine now.

**Comment by bernhardu** [comment]
Wow, that's great news, thank you very much.
The patch has landed in linux [v5.4.194](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.4.194), [v5.10.115](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.10.115), [v5.15.39](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.15.39), [v5.17.7](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/log/?h=v5.17.7) and [v5.18-rc6](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/?h=v5.18-rc6).
Now I am curious when it will hit [distribution](https://tracker.debian.org/pkg/linux) kernels.

**Comment by bernhardu** [comment]
I was too curious to wait, so I built a Debian kernel 5.16.18-1 with the mentioned patch.
And I can confirm the test suit succeeds the test suite for both, an amd64 VM and a i386 VM bookworm VM.
(host with zen_workaround.ko and in host and guest "echo 1 > /proc/sys/kernel/perf_event_paranoid")

So this issue should be closed then, as it never was really an issue with rr itself?
Thanks again.
